8251329: (zipfs) Files.walkFileTree walks infinitely if zip has dir named "." inside

jaikiran opened the pull request:
Can I please get a review of this proposed fix for https://bugs.openjdk.java.net/browse/JDK-8251329?

As noted in that issue, if a zip filesystem created on top of a jar containing a "./" entry is "walked" using the `Files.walkFileTree`, it leads to a infinite never ending iteration (which ultimately fails with Java heap space OOM).

Alan notes in that issue that:

> This is more likely an issue with the zipfs DirectoryStream implementation. A DirectoryStream is specified to not include elements that for the special links to the current or parent directory. It should be rare. 

This indeed turned out to be an issue in the `jdk.nio.zipfs.ZipDirectoryStream#iterator()` implementation where it calls the `jdk.nio.zipfs.ZipFileSystem#iteratorOf(...)` implementation.  The implementation, unlike what the javadoc of `java.nio.file.DirectoryStream` states, was including the "." and ".." paths in its returned iterator:

> The elements returned by the iterator are in no specific order. Some file
 systems maintain special links to the directory itself and the directory's
  parent directory. Entries representing these links are not returned by the
  iterator.


The proposed fix in this patch checks the paths for "." and "..", similar to what the `sun.nio.fs.UnixDirectoryStream` does in its `isSelfOrParent` and skips those paths from being added into the returned iterator. The `jdk.nio.zipfs.ZipFileSystem#iteratorOf(...)` (where this change has been done) is currently only used by `jdk.nio.zipfs.ZipDirectoryStream#iterator()` and has package-private visibility, so this change shouldn't impact any other usage/expectations.

A new jtreg test has been added to reproduce this issue and verify the fix. Local testing of the `test/jdk/jdk/nio/` (including this new test) went fine without any issues after this change. I will be triggering a `tier1` test locally in a while.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8251329](https://bugs.openjdk.java.net/browse/JDK-8251329): (zipfs) Files.walkFileTree walks infinitely if zip has dir named "." inside


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4604/head:pull/4604` \
`$ git checkout pull/4604`

Update a local copy of the PR: \
`$ git checkout pull/4604` \
`$ git pull https://git.openjdk.java.net/jdk pull/4604/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4604`

View PR using the GUI difftool: \
`$ git pr show -t 4604`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4604.diff">https://git.openjdk.java.net/jdk/pull/4604.diff</a>

</details>

AlanBateman commented on src/jdk.zipfs/share/classes/jdk/nio/zipfs/ZipFileSystem.java:
```diff
@@ -1957,6 +1961,15 @@ private OutputStream getOutputStream(Entry e) throws IOException {
+                    (nameAsBytes.length == 2 && nameAsBytes[1] == '.');
+        }
+        return false;
+    }
```
It might be a bit clearer if ZipPath define isSelfOrParent(), that would avoid needing to expose the internal representation.

jaikiran replied:
Done. I've updated the PR to implement this suggestion.

AlanBateman reviewed (commented):
The implementation change look good.

I skimmed the test. A helper method that maps a Set<Path> to a modifiable Set<String> would avoid the 10 usages of toString() in the setup, just a suggestion. Also I was puzzled by fs.getRootDirectories().iterator().next() and why this isn't fs.getPath("/").

jaikiran commented:
>I skimmed the test. A helper method that maps a Set to a modifiable Set would avoid the 10 usages of toString() in the setup, just a suggestion.

Done. I introduced a helper in the updated version of the PR.

> Also I was puzzled by fs.getRootDirectories().iterator().next() and why this isn't fs.getPath("/").

When I started with a reproducer and then this test, I used the code that the reporter had attached in the JBS issue. That one had used "fs.getRootDirectories().iterator().next()" to show this issue. I just happened to carry that over into the test. I've updated the PR to now use fs.getPath("/") and also verified that this version of the test continues to fail without the fix proposed in this PR and passes with the changes.

LanceAndersen commented:
Hi Jaikiran,

Consider:

```
        try (var os = Files.newOutputStream(ZIPFILE);
             ZipOutputStream zos = new ZipOutputStream(os)) {
            zos.putNextEntry(new ZipEntry("../Hello.txt"));
            zos.write("Hello World".getBytes(StandardCharsets.UTF_8));
        }
```

With your proposed fix, you will only return "/" when you walk the the Zip file and we should also return "/Hello.txt"

If. you resolve the path when the Inode entry is created:

           ` name(new ZipPath(null, normalize(name), true).getResolvedPath());`

That should address the issue and also allow you to access the entry.

AlanBateman commented:
I discussed this issue with @LanceAndersen and I think we are in agreement that a zip file system needs to use "." and ".." as links to the current and parent directories and cannot reliably support entries that have "." and ".." as name elements.  The other options on the table (including Lance's origin prototype fix and the proposal in this PR) lead to anomalies. So I think we should close this PR and restart JDK-8251329.  @jaikiran Are you okay this this? JDK-8251329 is still assigned to JDK-8251329 and has a patch with the changes that we think are the right way for newFileSystem to reject ZIP files that have entries in the CEN that can't be used for files in a file system.

jaikiran commented:
Hello Alan,

> So I think we should close this PR and restart JDK-8251329. @jaikiran Are you okay this this?

Yes, this sounds fine to me. I don't have any objection in closing this PR and I'll do it shortly.

> JDK-8251329 is still assigned to JDK-8251329 and has a patch with the changes that we think are the right way for newFileSystem to reject ZIP files 

This part I didn't understand. Did you mean to refer some other JBS issue? Because from what I see in https://bugs.openjdk.java.net/browse/JDK-8251329 there's no patch attached to it (unless of course it's restricted to specific JBS user roles?).

jaikiran commented:
Closing this PR based on the above discussion.

AlanBateman commented:
> This part I didn't understand. Did you mean to refer some other JBS issue? Because from what I see in https://bugs.openjdk.java.net/browse/JDK-8251329 there's no patch attached to it (unless of course it's restricted to specific JBS user roles?).

I wasn't suggesting there is a patch attached to that issue. Rather I was just pointing out that JDK-8251329 was being worked on already before this PR was created. Lance's initial patch for JDK-8251329 changed ZipPath::getResolvedPath when creating the inode entry but that approach leads to anomalies. Overall I think the discussion has been useful but I don't think it's feasible to have a file system view over entries that look like links to the current or parent directory.

jaikiran commented:
> I wasn't suggesting there is a patch attached to that issue. Rather I was just pointing out that JDK-8251329 was being worked on already before this PR was created.

Ok, that makes sense. Thank you for the details.
